File: algorithms/medium/nth_digit.cpp
class Solution {
public:
  int findNthDigit(int n) {
    long digit_num = 1;
    long ct = 9;
    while (digit_num * ct < n) {
      n -= digit_num * ct;
      digit_num++;
      ct *= 10;
    }

    int target = pow(10, (digit_num - 1)) + (n - 1) / digit_num;
    string s = to_string(target);
    return s[(n - 1) % digit_num] - '0';
  }
};

File: algorithms/medium/nth_digit.rs
impl Solution {
    pub fn find_nth_digit(n: i32) -> i32 {
        let mut n = n as u32;
        let mut digit_num = 1u32;
        let mut ct = 9u32;
        while digit_num * ct < n {
            n -= digit_num * ct;
            digit_num += 1;
            ct *= 10;
        }

        let target = 10u32.pow(digit_num - 1) + (n - 1) / digit_num;
        let s = format!("{}", target);
        let c = s.as_bytes()[((n - 1) % digit_num) as usize];
        (c - b'0') as _
    }
}
